### ISSUE TITLE
Fatal: hash does not match id

### ISSUE THREAD

**Issue by MichaelEischer** [restic/restic#3435]
_Originally posted by @malemburg in https://github.com/restic/restic/issues/1999#issuecomment-857046947_

I've compiled all the information requested in the template.

How can I run "debug examine" ?

Output of `restic version`
--------------------------

On the server:

restic 0.12.0 compiled with go1.15.8 on linux/amd64

On the notebook:

restic 0.12.0 compiled with go1.15.8 on windows/amd64 

How did you run restic exactly?
-------------------------------

On the notebook:

```
set RESTIC_REPOSITORY=\\<server>\restic
set RESTIC_PASSWORD_FILE=c:\restic\restic-repo-pwd.txt
set LOG=c:\restic\backup.log
c:\restic\restic backup --use-fs-snapshot C:\Users\<username> C:\Data >> %LOG% 2>&1
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------

A Linux server, running Samba 4.11. The repo is mounted on the notebook as an SMB share.

The notebook is running Window 10 20H2 with all patches applied.

Expected behavior
-----------------

Successful snapshot or at least an interrupted run without broken repo files (extra packs are fine, even unindexed ones).

Actual behavior
---------------

Every few weeks, backups stop to work. This often happens as a result of network connectivity issues of the notebook, which then result in the backup process taking too long and the Windows 10 scheduler terminating the process.

Here's a log extract from the notebook (with some parts translated from German):

```
...
Save(<data/eec07c2b38>) returned error, retrying after 15.321836548s: OpenFile: open \\?\UNC\<server>\restic\data\ee\eec07c2b380531c74fd3f5302ea22551d94f003d6d64af20515f3073c4f3766e: File exists
Save(<data/aecefbba94>) returned error, retrying after 26.801974529s: OpenFile: open \\?\UNC\<server>\restic\data\ae\aecefbba940d1fa1db1179cff5c90ef9bbfadc42c04cc9e234089972554a1141: File exists
Fatal: unable to save snapshot: OpenFile: open \\?\UNC\<server>\restic\data\29\29047f6b4348254ad25ce2ede71b60ebfe8df3ef66fd3d522a09754158beb534: File exists
```

The repo on the server ends up being full of non-indexed pack files as a result.

Rebuilding the index using rebuild-index works fine. This does not appear to check the hashes.

But analysis on the server using restic prune stops with an error, e.g.

```
...
repacking packs
[1:40:06] 46.52%  17169 / 36904 packs repacked
Fatal: hash does not match id: want acf2bbecc59239c0c2bbfe2769dc89f418a3d23f1904bc8a63f1469b6957352f, got 25f88ec3064875a900e910b83c5ee21e47934e8c3e652b6d13e4d12e4cb7ac0b
```

Checking on the server then confirms this:

```
# ll data/ac/acf2bbecc59239c0c2bbfe2769dc89f418a3d23f1904bc8a63f1469b6957352f
-rwxrw-r-- 1 x x 5177464 Jun  4 18:45 data/ac/acf2bbecc59239c0c2bbfe2769dc89f418a3d23f1904bc8a63f1469b6957352f
```

The file size roughly matches the file sizes of other data/ files.

```
# sha256sum data/ac/acf2bbecc59239c0c2bbfe2769dc89f418a3d23f1904bc8a63f1469b6957352f
25f88ec3064875a900e910b83c5ee21e47934e8c3e652b6d13e4d12e4cb7ac0b  data/ac/acf2bbecc59239c0c2bbfe2769dc89f418a3d23f1904bc8a63f1469b6957352f
```

In previous cases, where this happened, the files in question were clearly
truncated and had a much lower size than the other files under data/. However, not in this latest case. The file may still be truncated, but it's not as obvious as it was before.

Steps to reproduce the behavior
-------------------------------

I have not yet found a way to reproduce this. However, interrupting the backup process via the scheduler or Process Explorer does appear to trigger the issue.

Do you have any idea what may have caused this?
-----------------------------------------------

My guess is that the interaction of restic with the SMB mount is causing this. 

E.g. it's possible that restic is writing to the file on the notebook, this gets (write) cached by the OS and then fails to flush to the server after the process gets terminated, creating a truncated file on the repo server.

The server cannot be at fault, since this is used by several other systems without any problems.

The notebook certainly does have network connectivity issues due to poor network coverage at times, but otherwise works fine and does not exhibit crashes, RAM or CPU issues. I'll run Prime95 as extra measure.

Do you have an idea how to solve the issue?
-------------------------------------------

I don't know how restic works internally, but it is strange that files can end up in the repository which don't fulfill the invariant of sha256sum(file) == filename.

Restic should make sure that files which are written to the repo file system (even for network mounts) are consistent and only end up as valid repo files under data/ iff their hash matches the filename. If not already done, files should first be written to a temporary dir and then moved to data/ after successful completion only.

It may also be good to (possibly optionally) double check the written file by reading it back from the repo to check the hash and only add it to the data/ dir in case the hash matches.

**Comment by MichaelEischer** [comment]
_Originally posted by @cyounkins in https://github.com/restic/restic/issues/1999#issuecomment-857260116_

@malemburg thanks for the additional information!

If the SMB share is mounted locally, I believe restic would use the 'local' backend. Reading `internal/backend/local/local.go`, it appears that the file is created and written without a rename. It would be great for a developer with better understanding of the codebase to confirm this. Tools like `rsync` use a temp file when synchronizing, and I agree with @malemburg that doing similar could help preserve the invariant `sha256sum(file) == filename`.

If the connection were dropped, restic should be logging an error or something like `wrote %d bytes instead of the expected %d bytes`. Is something like that present? What does the very end of the log look like?

After looking at the code, I agree that if Windows is killing restic during backup it may leave truncated files. Maybe you can get Windows to not do that? :-)

```
Save(<data/eec07c2b38>) returned error, retrying after 15.321836548s: OpenFile: open \\?\UNC\<server>\restic\data\ee\eec07c2b380531c74fd3f5302ea22551d94f003d6d64af20515f3073c4f3766e: File exists
Save(<data/aecefbba94>) returned error, retrying after 26.801974529s: OpenFile: open \\?\UNC\<server>\restic\data\ae\aecefbba940d1fa1db1179cff5c90ef9bbfadc42c04cc9e234089972554a1141: File exists
```

This is odd to me because it seems to suggest that retries of these errors were successful when there's no reason they should be.

> "The server cannot be at fault, since this is used by several other systems without any problems."

restic seems to stress systems and detect more faults than many other programs. Also consider that Samba is very frequently configured in async mode for performance reasons, and it has the effect that writes are acknowledged before actually being written to disk. I haven't heard of an issue with that other than the case of power loss on the server, but it's possible. Do other people use the same samba server with restic?

**Comment by MichaelEischer** [comment]
> The repo on the server ends up being full of non-indexed pack files as a result.

I wouldn't recommend running `rebuild-index` in that case. `rebuild-index` is only intended for cases in which the repository was damaged and not for failed/interrupted backup runs. The non-indexed files would normally simply be removed by `prune` at a later time.

> Reading `internal/backend/local/local.go`, it appears that the file is created and written without a rename. It would be great for a developer with better understanding of the codebase to confirm this. Tools like `rsync` use a temp file when synchronizing, and I agree with @malemburg that doing similar could help preserve the invariant `sha256sum(file) == filename`.

Currently, the local backend in fact doesn't guarantee that pack files are written atomically. However, this is not really relevant for the integrity of the repository: the invariant `sha256sum(file) == filename` only has to hold for pack files which are referenced in the index and/or are actually used by a snapshot. restic strictly ensures that a pack file is first written successfully and flushed to disk using sync, before referencing blobs in that pack file from anywhere else. That is files which were not written completely (from the perspective of restic) won't be used. Putting things together this means that restic thought that the damaged pack file was written successfully, but somehow some parts of the file were broken/lost anyways.

Regarding SMB shares, I'm not sure whether the flush to disk part works or not. And I wonder which exact durability guarantees SMB gives after successfully writing and closing a file. I haven't been able to find much information on that topic. Rereading a file immediately after writing it, might not help as the correct file content could still be cached somewhere.

Since Samba 4.7 `strict sync = yes` seems to be the default, which makes it sound like `Sync` operations should be possible. On the other hand Microsoft has recently added a Write-Through option for SMB to increase the data storage reliability, which sounds like there we some cases in which this wasn't guaranteed...

What was the first error reported by `Save(<data/eec07c2b38>)` or `Save(<data/aecefbba94>)`? The later retries report that the file already exists which also means that restic failed to delete the incomplete pack file.

**Comment by malemburg** [comment]
Thanks for the note about rebuild-index. Sounds like it's actually better to only run prune regularly and use rebuild-index only in situations where restic flags this as a possible way to fix the repo.

FWIW: I have since switched from using local Samba mounts to SFTP and haven't seen similar repo corruptions since. I switched after I found that recent Windows 10 version do come with the client parts of OpenSSH readily pre-installed (the server parts can optionally be added as well), so that's a nice way to use restic without any additional client side installation requirements.

I would still suggest to make the local storage mechanism more reliable with an option to use temporary files before finalizing and adding the file to the index in the repo. Mounted network drives will always be vulnerable to issues with network connections dropping and backups should be as reliable as possible even under such conditions.
